### PR TITLE
Refactor/session data

### PR DESCRIPTION
**Pull request by alexanderleegs**
## Problem

This PR implements the first part of our new login flow. It refactors our existing method for passing request data from the router layer to services. Previously, we were decrypting the jwt stored in the isomercms cookie and storing it directly in `res.locals`, and each field was retrieved individually and passed down to services by the router. As part of our move to having multiple methods of login (github and email), we have opted to encapsulate all information related to a request in a new `SessionData` object. More information on the login flow refactor is contained in [this doc](https://docs.google.com/document/d/17RWEE-TZbk6PgM04mGOMPUM6UwFZYagF8POzwgh4S3k/edit#).

The following changes have been introduced in this PR:

- `SessionData` class which replaces the existing flow of passing data individually - this helps to facilitate future changes to the data being stored in the cookie, or other related session data. The related middleware and routes/services have also been modified accordingly
- Removed `whoamiAuth` middleware - it has been swapped out for the `verifyJwt` middleware, which has a similar purpose (decrypting and verification of user details)
- Replaced `useSiteTokenIfAvailable` middleware - we no longer plan to store a site access token for each site. The portion checking whether a user has access to a site has been extracted into a separate middleware (`checkHasAccess`) and used in its place.
- Tests have been updated to work with the new format. Auth middleware has also been refactored to typescript to get tests working.

These changes have been tested on staging with e2e testing completed. The email login specific changes will be implemented in a separate PR.

### CODE REVIEW

**Review by seaerchin — COMMENTED** [review]
it feels like we're passing `sessionData` through everything but most services only require an extremely limited subset of its functionality, which is not clear until we trace the method call all the way.

this pattern feels dangerous because `sessionData` is temporary and we will need to delete it in the future and js doesn't tell us whether we still have the necessary properties (ie, we have to manually trace and verify). 

we might wanna shift the extraction of props from `sessionData` higher up rather than retaining it inside `GithubService` just to make this prop extraction clear. this also applies to methods taking `accessToken` previously now taking `sessionData`

Another problematic issue is that `SessionData` now has **mutable** internal states - if we have middleware setup before, then the `GithubState` is set (and similarly so for `siteName`). this feeds into the aforementioned difficulty of telling what is used from `sessionData` without tracing and also, necessitates an implicit link between the middlewares and `sessionData` (ie, calling `getGithubState/getSiteName` without the middleware is a bug) - i'm not too sure on how the required props functioned before (it seemed like they were just set on `res.locals` and passed around) but i'm not too keen on having this. sadly, i also am p unsure of how to avoid this beyond just passing the required props directly to the method call rather than having an intermediary. 

wdyt?

**Comment by seaerchin** [comment]
as an aside, would it be possible for future PRs of this scale to be split into independent pieces (eg: introduction of class/refactor/fix tests) for ease of review?

i felt p unsure of what to focus on while reviewing due to the sheer # of files changed .-.

**Comment by alexanderleegs** [comment]
> it feels like we're passing `sessionData` through everything but most services only require an extremely limited subset of its functionality, which is not clear until we trace the method call all the way.
> 
> this pattern feels dangerous because `sessionData` is temporary and we will need to delete it in the future and js doesn't tell us whether we still have the necessary properties (ie, we have to manually trace and verify).
> 
> we might wanna shift the extraction of props from `sessionData` higher up rather than retaining it inside `GithubService` just to make this prop extraction clear. this also applies to methods taking `accessToken` previously now taking `sessionData`
> 
> Another problematic issue is that `SessionData` now has **mutable** internal states - if we have middleware setup before, then the `GithubState` is set (and similarly so for `siteName`). this feeds into the aforementioned difficulty of telling what is used from `sessionData` without tracing and also, necessitates an implicit link between the middlewares and `sessionData` (ie, calling `getGithubState/getSiteName` without the middleware is a bug) - i'm not too sure on how the required props functioned before (it seemed like they were just set on `res.locals` and passed around) but i'm not too keen on having this. sadly, i also am p unsure of how to avoid this beyond just passing the required props directly to the method call rather than having an intermediary.
> 
> wdyt?

As discussed offline, there's a concern that `SessionData` is growing into a god object equivalent that future engineers will be tempted to insert any state into in future, which is something we want to avoid. However, the `SessionData` concept is still useful overall, as it helps us to transition between github and email logins, and can be useful in future for differentiation based on data stored.

As a compromise, we have decided to remove the mutable states in favour of splitting the `SessionData` class into several different classes with different responsibilities - this helps to discourage the dumping of state into a single class and also allows us to identify the data required by a method more easily. We will also continue passing `SessionData` as this is required by future PRs.

File: src/classes/SessionData.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
+export interface SessionDataProps {
+  githubId?: string
+  accessToken?: string
+  isomerUserId: string
+  email?: string
+}
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/classes/SessionData.ts]: Good catch, fixed in https://github.com/isomerpages/isomercms-backend/pull/477/commits/52a386c771df1c58f2db6107f8a6e5368ceac2f7

File: src/classes/SessionData.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    return this.isomerUserId
+  }
+
+  getEmail() {
+    return this.email
+  }
+
+  getIsEmailUser() {
+    return this.isEmailUser
+  }
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/classes/SessionData.ts]: We're planning to keep this class even after moving off github login completely - but you're right about the pure retrieval methods not being very useful. They've been removed in https://github.com/isomerpages/isomercms-backend/pull/477/commits/2f7175977e2fe0662a3366b1689b117016bae864!

File: src/classes/SessionData.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  getGithubState() {
+    return {
+      currentCommitSha: this.currentCommitSha,
+      treeSha: this.treeSha,
+    }
+  }
+
+  addSiteName(siteName: string) {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/classes/SessionData.ts]: Removed mutable state in https://github.com/isomerpages/isomercms-backend/pull/477/commits/2f7175977e2fe0662a3366b1689b117016bae864

File: src/classes/SessionData.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  getEmail() {
+    return this.email
+  }
+
+  getIsEmailUser() {
+    return this.isEmailUser
+  }
+
+  addGithubState({
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/classes/SessionData.ts]: Removed mutable state in https://github.com/isomerpages/isomercms-backend/pull/477/commits/2f7175977e2fe0662a3366b1689b117016bae864

File: src/middleware/index.ts [hunk 5]
```diff
@@ -40,4 +45,13 @@ const attachFormSGHandler = (formKey: string): RequestHandler[] => [
   formSGService.decrypt({ formKey }),
 ]
 
-export { getAuthMiddleware, attachFormSGHandler }
+const attachSiteHandler = (req: Request, res: Response, next: NextFunction) => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/middleware/index.ts]: Makes sense, fixed in https://github.com/isomerpages/isomercms-backend/pull/477/commits/bbdb6a23de1747b98678e8e7c52f7190340373e5

File: src/services/configServices/NetlifyTomlService.js [hunk 6]
```diff
@@ -12,7 +12,8 @@ const {
 const { GITHUB_BUILD_ORG_NAME, GITHUB_BUILD_REPO_NAME } = process.env
 
 class NetlifyTomlService {
-  async read({ accessToken }) {
+  async read(sessionData) {
+    const accessToken = sessionData.getAccessToken()
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/configServices/NetlifyTomlService.js]: As discussed offline, we're leaving this as is!

File: src/services/configServices/SettingsService.js [hunk 7]
```diff
@@ -17,12 +17,12 @@ class SettingsService {
...
+  async retrieveSettingsFiles(sessionData, shouldRetrieveHomepage) {
     const fileRetrievalObj = {
-      config: this.configYmlService.read(reqDetails),
-      footer: this.footerYmlService.read(reqDetails),
-      navigation: this.navYmlService.read(reqDetails),
-      homepage: this.homepagePageService.read(reqDetails),
+      config: this.configYmlService.read(sessionData),
+      footer: this.footerYmlService.read(sessionData),
+      navigation: this.navYmlService.read(sessionData),
+      homepage: this.homepagePageService.read(sessionData),
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/configServices/SettingsService.js:25]: See above

File: src/services/db/GitHubService.js [hunk 8]
```diff
@@ -69,7 +71,9 @@ class GitHubService {
     }
   }
 
-  async read({ accessToken, siteName }, { fileName, directoryName }) {
+  async read(sessionData, { fileName, directoryName }) {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/db/GitHubService.js:74]: See above

File: src/services/directoryServices/CollectionDirectoryService.js [hunk 9]
```diff
@@ -85,8 +85,8 @@ class CollectionDirectoryService {
     return fileOrder
   }
 
-  async listAllCollections(reqDetails) {
-    const filesOrDirs = await this.baseDirectoryService.list(reqDetails, {
+  async listAllCollections(sessionData) {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/directoryServices/CollectionDirectoryService.js:88]: See above

File: src/services/directoryServices/MediaDirectoryService.js [hunk 10]
```diff
@@ -30,16 +30,16 @@ class MediaDirectoryService {
     return files
   }
 
-  async listFiles(reqDetails, { directoryName }) {
+  async listFiles(sessionData, { directoryName }) {
     // TODO: file preview handling
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/directoryServices/MediaDirectoryService.js]: Removed in https://github.com/isomerpages/isomercms-backend/pull/477/commits/942b0058a435cbd46e7fdf2ab934995c28927302 as it seems to be already resolved

File: src/services/directoryServices/MediaDirectoryService.js [hunk 11]
```diff
@@ -30,16 +30,16 @@ class MediaDirectoryService {
...
-    const { siteName } = reqDetails
+    const { siteName } = sessionData
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/directoryServices/MediaDirectoryService.js:34]: Yeah there doesn't seem to be an issue

File: src/services/utilServices/AuthService.js [hunk 12]
```diff
@@ -84,12 +84,14 @@ class AuthService {
...
       isomer_user_id: user.id,
+      email: user.email,
     })
 
     return token
   }
 
-  async getUserInfo({ accessToken }) {
+  async getUserInfo(sessionData) {
+    const accessToken = sessionData.getAccessToken()
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/services/utilServices/AuthService.js]: See above!

**Review by seaerchin — APPROVED** [review]
lgtm; have comment wrt simplifying some code but rest is fine. we do have to take care to ensure that the threading of the object doesn't propagate throughout our codebase though but it's a choice we'll have to live w/ 

also, due to the huge # of files changed, could i trouble you to run unit tests prior to merge (ideally also e2e).

File: src/classes/UserWithSiteSessionData.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+  getGithubParamsWithSite() {
+    return {
+      githubId: this.githubId,
+      accessToken: this.accessToken,
+      isomerUserId: this.isomerUserId,
+      email: this.email,
+      siteName: this.siteName,
+    }
+  }
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/classes/UserWithSiteSessionData.ts]: Opting to leave the distinction in so that it's clearer when we need the sitename!

**Pull request merged**